### PR TITLE
Sites and sheaves

### PR DESCRIPTION
**Pull request by plt-amy**
Stuff from C2.1 in the elephant

### CODE REVIEW

File: src/Cat/Diagram/Sieve.lagda.md [hunk 1]
```diff
@@ -109,8 +126,32 @@ proposition). Since natural transformations are monic if they are
...
 ```agda
-  to-presheaf↪よ : {S : Sieve C c} → to-presheaf S PSh.↪ よ₀ C c
+  to-presheaf↪よ : ∀ {c} {S : Sieve C c} → to-presheaf S PSh.↪ よ₀ C c
   to-presheaf↪よ {S} .mor .η x (f , _) = f
   to-presheaf↪よ {S} .mor .is-natural x y f = refl
   to-presheaf↪よ {S} .monic g h path = ext λ i x → Σ-prop-path! (unext path i x)
 ```
+
+## Pullback of sieves
+
```
Comment (plt-amy) [inline src/Cat/Diagram/Sieve.lagda.md:136]: This needs prose

File: src/Cat/Diagram/Sieve.lagda.md [hunk 2]
```diff
@@ -109,8 +126,32 @@ proposition). Since natural transformations are monic if they are
...
+  pullback-id : ∀ {c} {s : Sieve C c} → pullback C.id s ≡ s
+  pullback-id {s = s} = ext λ h → Ω-ua (subst (_∈ s) (C.idl h)) (subst (_∈ s) (sym (C.idl h)))
+
+  pullback-∘
+    : ∀ {u v w} {f : C.Hom w v} {g : C.Hom v u} {R : Sieve C u}
+    → pullback (g C.∘ f) R ≡ pullback f (pullback g R)
+  pullback-∘ {f = f} {g} {R = R} = ext λ h →
+    Ω-ua (subst (_∈ R) (sym (C.assoc g f h))) (subst (_∈ R) (C.assoc g f h))
+
+  Sieves : Functor (C ^op) (Sets (o ⊔ ℓ))
```
Comment (plt-amy) [inline src/Cat/Diagram/Sieve.lagda.md:151]: This is the subobject classifier in PSh(C) and the prose should mention that

File: src/Cat/Functor/Reasoning/Presheaf.agda [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+expand p = happly (Pf.expand p) _
+
+weave : f ∘ g ≡ h ∘ i → F₁ f (F₁ g x) ≡ F₁ h (F₁ i x)
+weave p = happly (Pf.weave p) _
+
+annihilate : f ∘ g ≡ id → F₁ f (F₁ g x) ≡ x
+annihilate p = happly (Pf.annihilate p) _
+
+conjure : id ≡ f ∘ g → x ≡ F₁ f (F₁ g x)
+conjure p = sym (annihilate (sym p))
```
Comment (plt-amy) [inline src/Cat/Functor/Reasoning/Presheaf.agda:49]: I actually didn't use most of these but they're sure to be handy

File: src/Cat/Functor/Reasoning/Presheaf.agda [hunk 4]
```diff
@@ -0,0 +1,49 @@
+open import Cat.Prelude
+
+import Cat.Functor.Reasoning as Func
+import Cat.Reasoning as Cat
+
+module Cat.Functor.Reasoning.Presheaf
+  {o ℓ s} {C : Precategory o ℓ} (P : Functor C (Sets s))
```
Comment (plt-amy) [inline src/Cat/Functor/Reasoning/Presheaf.agda:7]: Module is a slight misnomer but not having the `^op` lets Agda actually solve for `C`, so it's a win

File: src/Cat/Site/Base.lagda.md [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+<!--
+```agda
+private variable
+  o ℓ ℓc ℓs : Level
+  C : Precategory o ℓ
+```
+-->
+
+# Sites and sheaves
+
```
Comment (plt-amy) [inline src/Cat/Site/Base.lagda.md:25]: Prose

File: src/Cat/Site/Base.lagda.md [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+```
+-->
+
+```agda
+  record Patch {U} (T : Sieve C U) : Type (o ⊔ ℓ ⊔ ℓs) where
+    no-eta-equality
+    field
+      part  : Parts T
+      patch : is-patch T part
+```
```
Comment (plt-amy) [inline src/Cat/Site/Base.lagda.md:101]: This is what Johnstone calls a "compatible family of elements", but that's way too wordy to use in a type name, so I invented this

File: src/Cat/Site/Base.lagda.md [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+
+## Separated presheaves and sheaves
+
+<!--
+```agda
+module _ {o ℓ ℓs} {C : Precategory o ℓ} (P : Functor (C ^op) (Sets ℓs)) where
+  open Precategory C
+```
+-->
+
```
Comment (plt-amy) [inline src/Cat/Site/Base.lagda.md:203]: Prose

File: src/Cat/Site/Base.lagda.md [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+```agda
+  is-separated₁ : ∀ {U} (T : Sieve C U) → Type _
+  is-separated₁ {U} T =
+    ∀ {x y : P ʻ U}
+    → (l : ∀ {V} (f : Hom V U) (hf : f ∈ T) → P ⟪ f ⟫ x ≡ P ⟪ f ⟫ y)
+    → x ≡ y
```
Comment (plt-amy) [inline src/Cat/Site/Base.lagda.md:209]: This is much better to use than `(f : Patch P T) → is-prop (Section P f)`, but is equivalent to it.

File: src/Cat/Site/Base.lagda.md [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+  record is-sheaf : Type (o ⊔ ℓ ⊔ ℓs ⊔ ℓc) where
+    field
+      has-sheaf₁ : ∀ {U} (c : J .covers U) → is-sheaf₁ P (J .cover c)
+
+    split : ∀ {U : ⌞ C ⌟} {c : J # U} (p : Patch P (J .cover c)) → Section P p
+    split {c = c} p = has-sheaf₁ c p .centre
+
+    abstract
+      separate : ∀ {U : ⌞ C ⌟} (c : J # U) → is-separated₁ P (J .cover c)
+      separate c l = is-sheaf₁→is-separated₁ P (J .cover c) (has-sheaf₁ c) l
```
Comment (plt-amy) [inline src/Cat/Site/Base.lagda.md:239]: Having this as a record type allows solving for the coverage `J`

File: src/Cat/Site/Base.lagda.md [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+```agda
+  Sheaves : Precategory (o ⊔ ℓ ⊔ ℓc ⊔ lsuc ℓs) (o ⊔ ℓ ⊔ ℓs)
+  Sheaves .Ob = Σ[ p ∈ Functor (C ^op) (Sets ℓs) ] is-sheaf J p
+  Sheaves .Hom (S , _) (T , _) = S => T
+  Sheaves .Hom-set _ _ = hlevel 2
+  Sheaves .id = idnt
+  Sheaves ._∘_ = _∘nt_
+  Sheaves .idr f = trivial!
+  Sheaves .idl f = trivial!
+  Sheaves .assoc f g h = trivial!
```
Comment (plt-amy) [inline src/Cat/Site/Base.lagda.md:269]: This prints better than `Restrict (Functor (C ^op) (Sets ℓs))  (is-sheaf J)` but I could use that instead

File: src/Cat/Site/Closure.lagda.md [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+import Cat.Functor.Reasoning.Presheaf as Psh
+import Cat.Reasoning as Cat
+```
+-->
+
+```agda
+module Cat.Site.Closure where
+```
+
+# Closure properties of sites
```
Comment (plt-amy) [inline src/Cat/Site/Closure.lagda.md:21]: Actually it's "closure properties of the sieves that sheaves on a site are sheaves for" but that's extremely verbose

File: src/Cat/Site/Closure.lagda.md [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+```
+
+<!--
+```agda
+module _ {o ℓ ℓs} {C : Precategory o ℓ} (P : Functor (C ^op) (Sets ℓs)) where
+  open Precategory C
+```
+-->
+
+## The canonical topology
```
Comment (plt-amy) [inline src/Cat/Site/Closure.lagda.md:98]: This could probably be in another module

File: src/Cat/Site/Closure.lagda.md [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+
+## In general
+
+```agda
+  sieve→cocone : ∀ {U} (S : Sieve C U) → πₚ C (to-presheaf S) => Const U
+  sieve→cocone S .η x = x .section .fst
+  sieve→cocone S .is-natural x y f = ap fst (f .commute) ∙ introl refl
+
+  -- C2.2.12
+  is-effective-epi : ∀ {U} → Sieve C U → Type _
```
Comment (plt-amy) [inline src/Cat/Site/Closure.lagda.md:131]: This name is fine

File: src/Cat/Site/Closure.lagda.md [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+  is-effective-epi {U} S = is-colimit _ U (sieve→cocone S)
+
+  is-universally-effective-epi : ∀ {U} → Sieve C U → Type _
```
Comment (plt-amy) [inline src/Cat/Site/Closure.lagda.md:134]: This name is pushing it

File: src/Cat/Site/Closure.lagda.md [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+  is-universally-effective-epi {U} S =
+    ∀ {V} (f : Hom V U) → is-effective-epi (pullback f S)
+
+  is-uee→is-sheaf₁
```
Comment (plt-amy) [inline src/Cat/Site/Closure.lagda.md:138]: and this name is an actual trainwreck.

File: src/Cat/Site/Closure.lagda.md [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+  Canonical-coverage : Coverage C (o ⊔ ℓ)
+  Canonical-coverage .covers U = Σ[ s ∈ Sieve C U ] is-universally-effective-epi s
+  Canonical-coverage .cover = fst
+  Canonical-coverage .stable (c , st) f = pure $
+    (pullback f c , λ g → subst is-effective-epi pullback-∘ (st (f ∘ g))) , λ g x → x
+
+  representable-is-sheaf-canonical : ∀ {U} → is-sheaf Canonical-coverage (よ₀ C U)
+  representable-is-sheaf-canonical .has-sheaf₁ (c , st) = is-uee→is-sheaf₁ c st
+
+  is-sheaf₁→is-uee
```
Comment (plt-amy) [inline src/Cat/Site/Closure.lagda.md:174]: This result is weaker than what Johnstone promises just before C2.1.12 but I think it's the strongest possible; c.f. Zhen Lin's counterexample and correction to C2.1.7 [here](https://math.stackexchange.com/questions/358570/elephant-how-do-i-prove-lemma-2-1-7-section-c2-1/358709#358709). In particular we need stability under pullbacks which is exactly 2.1.7.

File: src/Cat/Site/Instances/Frame.lagda.md [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+```
+
+<!--
+```agda
+open Frame isf
+```
+-->
+
+# Frames as sites
+
```
Comment (plt-amy) [inline src/Cat/Site/Instances/Frame.lagda.md:27]: Prose

File: src/Cat/Site/Closure.lagda.md (hunk 11 above)
Comment (plt-amy) [reply src/Cat/Site/Closure.lagda.md:21]: Also prose

**Comment by Lavenza** [comment]
[Pull request preview](https://preview.1lab.dev/369)

File: src/Cat/Site/Sheafification.lagda.md [hunk 18]
```diff
@@ -0,0 +1,206 @@
...
+import Cat.Functor.Reasoning.Presheaf as Psh
+```
+-->
+
+```agda
+module Cat.Site.Sheafification where
+```
+
+# Sheafification
+
```
Comment (plt-amy) [inline src/Cat/Site/Sheafification.lagda.md:17]: Prose

File: src/Cat/Site/Sheafification.lagda.md [hunk 19]
```diff
@@ -0,0 +1,206 @@
...
+      → (g : pre.is-patch C map (J .cover c) p)
+      → Sheafify₀ U
+
+    glues
+      : (c : J .covers U) (p : pre.Parts C map (J .cover c))
+      → (g : pre.is-patch C map (J .cover c) p)
+      → pre.is-section C map (J .cover c) (glue c p g) p
+
+    squash : is-set (Sheafify₀ U)
+```
```
Comment (plt-amy) [inline src/Cat/Site/Sheafification.lagda.md:66]: I'll probably break this definition into tons of paragraphs

File: src/Cat/Site/Sheafification.lagda.md [hunk 20]
```diff
@@ -0,0 +1,206 @@
...
+
+  Sheafify-is-sheaf : is-sheaf J Sheafify
+  Sheafify-is-sheaf = from-is-separated J Sheafify Sheafify-is-sep λ c s → record
+    { part  = glue c (s .part) (s .patch)
+    ; patch = glues c (s .part) (s .patch)
+    }
+```
+
+## An induction principle
+
```
Comment (plt-amy) [inline src/Cat/Site/Sheafification.lagda.md:87]: We can't sweep this induction principle under the rug because it's actually quite different from the one you'd get automatically

Comment (plt-amy) [reply src/Cat/Site/Sheafification.lagda.md:87]: Also gotta make sure to credit @MatthiasHu and the cubical library. [code at time of writing](https://github.com/agda/cubical/blob/cabdab89e03d727419ce28c8b988ca36d89126fd/Cubical/Categories/Site/Sheafification/ElimProp.agda) [commit which added it](https://github.com/agda/cubical/commit/bd6f5dee80aec04c72452c99a3a0a39c94cf4b73)

File: src/Cat/Site/Sheafification.lagda.md [hunk 21]
```diff
@@ -0,0 +1,206 @@
...
+```agda
+module Small {ℓ} {C : Precategory ℓ ℓ} (J : Coverage C ℓ)  where
+  open Sheafification J
+
+  open _=>_
+
+  private variable F : Functor (C ^op) (Sets ℓ)
+```
+-->
+
```
Comment (plt-amy) [inline src/Cat/Site/Sheafification.lagda.md:157]: This needs a warning about the size restrictions

**Review by plt-amy — COMMENTED** [review]
Just leaving some notes to myself as review comments

File: src/Cat/Site/Closure.lagda.md (hunk 15 above)
Comment (ncfavier) [reply src/Cat/Site/Closure.lagda.md:138]: uee

**Comment by plt-amy** [comment]
Prose for Cat.Site.Base probably sucks, @ncfavier @TOTBWF could you review?

File: src/Cat/Site/Sheafification.lagda.md [hunk 22]
```diff
@@ -0,0 +1,403 @@
...
+    go (inc-natural {f = f} x i) = is-prop→pathp (λ i → p'prp (inc-natural x i)) (p'inc (A.₁ f x)) (p'map f (inc x) (p'inc x)) i
+    go (sep {x = x} {y = y} c l i) = is-prop→pathp (λ i → p'prp (sep c l i)) (go x) (go y) i
+    go (map-id x i) = is-prop→pathp (λ i → p'prp (map-id x i)) (p'map id x (go x)) (go x) i
+    go (map-∘ {g = g} {f = f} x i) = is-prop→pathp (λ i → p'prp (map-∘ x i)) (p'map (g ∘ f) x (go x)) (p'map f (map g x) (p'map g x (go x))) i
+    go (glues c p g f hf i) = is-prop→pathp (λ i → p'prp (glues c p g f hf i)) (p'map f (glue c p g) (p'glue c p g λ g hg → go (p g hg))) (go (p f hf)) i
+    go (squash x y p q i j) = is-prop→squarep (λ i j → p'prp (squash x y p q i j)) (λ i → go x) (λ i → go (p i)) (λ i → go (q i)) (λ i → go y) i j
+```
+-->
+
+## The universal property
```
Comment (ncfavier) [inline src/Cat/Site/Sheafification.lagda.md:284]: Can we package this up nicely as a reflective subcategory or something?

**Review by ncfavier — COMMENTED** [review]
Very nice work! I suppose the Elephant page can be updated?

File: src/Cat/Site/Base.lagda.md [hunk 23]
```diff
@@ -0,0 +1,578 @@
...
+  is-section {T = T} p x = pre.is-section C A.₁ T p (x .part)
+```
+-->
+
+```agda
+  record Section {U} {T : Sieve C U} (p : Patch T) : Type (o ⊔ ℓ ⊔ ℓs) where
+    no-eta-equality
+    field
+      {part} : A ʻ U
+      patch  : is-section part p
```
[the remark this answers is not in the document]
Comment (plt-amy) [reply src/Cat/Site/Base.lagda.md:227]: Do you have better suggestions?

File: src/Cat/Site/Sheafification.lagda.md (hunk 22 above)
Comment (plt-amy) [reply src/Cat/Site/Sheafification.lagda.md:284]: [it is](https://github.com/plt-amy/1lab/pull/369/files/8bc38b6da3167ab21aee44e50b6ef5c7d7dc6fae#diff-5b35c30b947586e0812ced49f8ca17571df85189dff18785e38b174f0eac6d4bR71)?

File: src/Cat/Site/Base.lagda.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (ncfavier) [reply src/Cat/Site/Base.lagda.md:227]: `section` and `has-is-section`?

[the remark this answers is not in the document]
Comment (plt-amy) [reply src/Cat/Site/Base.lagda.md:227]: We actually use that projection a lot though. That sucks

File: src/Cat/Site/Sheafification.lagda.md (hunk 22 above)
Comment (ncfavier) [reply src/Cat/Site/Sheafification.lagda.md:284]: Ah, I missed that. Can we split it into reflective + monadic?

File: src/Cat/Site/Base.lagda.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (plt-amy) [reply src/Cat/Site/Base.lagda.md:227]: Like all the has-is-X projections basically go unused because we reexport the relevant fields (which have much shorter names). I think part and patch make sense...

[the remark this answers is not in the document]
Comment (ncfavier) [reply src/Cat/Site/Base.lagda.md:227]: I guess in both cases the "part" is a (collection of) elements and the "patch" condition says that it agrees with something (itself, or another patch)? Still I think reusing the same field names is confusing.

nlab seems to say *matching family* for patch and *amalgamation* for section...

[the remark this answers is not in the document]
Comment (ncfavier) [reply src/Cat/Site/Base.lagda.md:227]: Wikipedia:

> The section $$s$$ whose existence is guaranteed by axiom 2 is called the *gluing*, *concatenation*, or *collation* of the sections $$s_i$$.

[the remark this answers is not in the document]
Comment (plt-amy) [reply src/Cat/Site/Base.lagda.md:227]: Right, a part is a thing and patch is a verb. And, yes, but Matching-family is also a shit name; I complained about this on mastodon months ago

[the remark this answers is not in the document]
Comment (ncfavier) [reply src/Cat/Site/Base.lagda.md:227]: I think part/patch are good names (or at least they're as good as names can ever be), I'm only objecting to reusing them for the gluing. Maybe `gluing` and `glues`?

[the remark this answers is not in the document]
Comment (plt-amy) [reply src/Cat/Site/Base.lagda.md:227]: The projections are entirely analogous though? Here is the actual thing, here's the proof it agrees with what it should. I don't understand why they should have different names

File: src/Cat/Functor/Hom/Yoneda.lagda.md [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+transformations into $A$, with domain a $\hom$-functor; and that this
+isomorphism is natural. This result is actually extremely easy to prove,
+as long as we expand *what* a natural transformation $\eta : \cC(-, U)
+\To A$ consists of: a dependent function
+$$\eta : \prod_{V : \cC} \cC(V, U) \to A(V)$$.
+
+There's no secret in choosing the components of the isomorphism:
+
+- Given a $x : A(U)$, we define a natural transformation
+  $\operatorname{yo} x$ by sending a map $h : \cC(V, U)$ to the
```
[the remark this answers is not in the document]
Comment (ncfavier) [reply src/Cat/Functor/Hom/Yoneda.lagda.md:39]: Oh, I should have read more carefully. Yeah that's fine then

File: src/Cat/Instances/Sheaves/Exponentials.lagda.md [hunk 25]
```diff
@@ -0,0 +1,141 @@
...
+
+What we must *show*, however, is that $x(f,z) = y(f,z)$, for a totally
+arbitrary $f : W \to V$ and $z : A(V)$. But, since this is an equality
+in $B(W)$, and $B$ is a $J$-sheaf, it suffices to choose a $J$-covering
+sieve $S$ of $W$, and to show this equality $S$-locally, i.e. to show
+$$
+B(g)(x(f, z)) = B(g)(y(f, z))
+$$
+for any $g \in S$. Note that, by naturality, these are equal to $x(fg,
+z)$,resp. $y(fg, z)$. We can then choose $S = f^*(R)$, which lets us
```
Comment (ncfavier) [inline src/Cat/Instances/Sheaves/Exponentials.lagda.md:72]:
```suggestion
z)$, resp. $y(fg, z)$. We can then choose $S = f^*(R)$, which lets us
```

File: src/Cat/Instances/Sheaves/Exponentials.lagda.md [hunk 26]
```diff
@@ -0,0 +1,141 @@
...
+between local equality of natural transformations and local equality of
+their components, by passing to a pullback. The rest of the proof
+essentially follows from similar calculations.
+
+This will also be the strategy to show that $B^A$ admits sections!  To
+put together a section for a [[patch]] $p$ for a $J$-covering sieve $R$,
+it suffices to do so at each component; so we'll start by showing that
+$p$ descends to a patch $p'$ for $B(U)$ on $f^*R$, for any $f : V \to
+U$, as long as we're given $x : A(V)$. Each part of $p'$ is given by
+evaluating the correspoding part of $p$:
```
Comment (ncfavier) [inline src/Cat/Instances/Sheaves/Exponentials.lagda.md:98]:
```suggestion
evaluating the corresponding part of $p$:
```

**Review by ncfavier — APPROVED** [review]
